### PR TITLE
fix(chat): preserve JSON completion semantics in streaming delivery

### LINKED ISSUE #3770: [Bug]: Chat JSON-to-SSE fallback drops tool calls/reasoning and overwrites finish_reason with stop
### Resolved in dev

[#3803](https://github.com/lidge-jun/opencodex/pull/3803) (`ac4a7659fd`) preserves tools, reasoning, usage and finish semantics in JSON-to-SSE completion delivery.

Verified against dev `5759d9ea2f1e7281cdc01eb9628f2e0a123fb59c`. Original report by @turin-dev. The attribution record was added in [#3811](https://github.com/lidge-jun/opencodex/pull/3811).

### Client or integration

Direct HTTP/API client

### Area

Streaming

### Summary

When the Chat endpoint's internal Responses replay returns JSON instead of SSE, `stream: true` enters a lossy JSON-to-SSE fallback. The fallback serializes only `message.content` and hard-codes `finish_reason: "stop"`, even though `responsesJsonToChatCompletion` has already produced tool calls, reasoning content, and the correct finish reason.

The same upstream JSON produces:
- `stream: false`: the expected `message.tool_calls` and `finish_reason: "tool_calls"`.
- `stream: true`: no tool-call delta at all; just the assistant role and a successful-looking `stop` followed by `[DONE]`.

A reasoning-plus-text response loses its `reasoning_content` while retaining normal text. A response incomplete due to `max_output_tokens` changes from `length` to `stop`.

Impact: tool-capable clients cannot execute the requested tool, and clients cannot detect token-limit truncation from the terminal chunk. This is specific to the JSON response fallback, not a claim about the usual SSE-to-SSE or native Chat passthrough paths.

Expected: the synthesized stream preserves the semantic fields and terminal reason of the already-converted Chat completion: indexed `delta.tool_calls`, `delta.reasoning_content`, and `choices[0].finish_reason`.

### Reproduction

Executed the unmodified `handleChatCompletions` and translation modules from commit `a349b521bb02f76c7b8820c13633a8076dcec378` with a mocked `handleResponses` boundary that returns `application/json`. Routing and unrelated runtime integrations are stubbed so the test needs no provider, credentials, or running proxy.

Save the following as `repro-json-chat.ts` in the repository root, then run it in a **separate process** with `bun run repro-json-chat.ts` (the mocks are process-global):

```ts
import { mock } from "bun:test";
let upstreamBody: Record<string, unknown>;
mock.module("./src/adapters/openai-responses", () => ({ FORWARD_HEADERS: [] }));
mock.module("./src/lib/retry-after", () => ({ resolveClientRetryAfter: () => undefined }));
mock.module("./src/lib/token-estimate", () => ({ estimateTokens: () => 1 }));
mock.module("./src/router", () => ({
  UnknownRoutingPolicyError: class extends Error {},
  NoEligiblePolicyCandidateError: class extends Error {},
  routeModel: () => ({
    providerName: "fixture", modelId: "fixture", codexAccountMode: "direct",
    provider: { adapter: "openai-responses" },
  }),
}));
mock.module("./src/routing/request-evidence", () => ({ evidenceFromBody: () => ({}) }));
mock.module("./src/server/adapter-resolve", () => ({ resolveWireProtocolOverride: (_name, _model, provider) => provider }));
mock.module("./src/server/request-decompress", () => ({ readJsonRequestBody: (req) => req.json() }));
mock.module("./src/server/request-log", () => ({
  addFinalRequestLog() {}, httpStatusForRequestLogTerminal: () => 200, recordFirstOutput() {},
}));
mock.module("./src/server/relay", () => ({ responseWithDeferredRequestLog: (r) => r }));
mock.module("./src/server/responses", () => ({ handleResponses: async () => Response.json(upstreamBody) }));
mock.module("./src/codex/native-main-admission", () => ({ tryClaimNativeMainProfileForTurn: () => false }));
mock.module("./src/server/chat-native", () => ({ isNativeChatRouteEligible: () => false, handleNativeChatCompletions() { throw Error("unexpected native path"); } }));
mock.module("./src/server/effort-row", () => ({ parseRequestEffortRowId: () => null }));
mock.module("./src/server/fast-row", () => ({ parseSyntheticRowId: () => ({}) }));
mock.module("./src/providers/openai-tiers", () => ({ isCanonicalOpenAiForwardProvider: () => false }));
mock.module("./src/codex/loopback-target", () => ({ CODEX_RESERVE_HELPER_UNSUPPORTED_MESSAGE: "", isCodexReserveHelperUnsupported: () => false }));
const { handleChatCompletions } = await import("./src/server/chat-completions");
const fixtures = {
  tool: { status: "completed", output: [{ type: "function_call", call_id: "call_fixture", name: "lookup", arguments: '{"q":"hello"}' }] },
  reasoning: { status: "completed", output: [{ type: "reasoning", summary: [{ type: "summary_text", text: "Fixture reasoning." }] }, { type: "message", role: "assistant", content: [{ type: "output_text", text: "Answer." }] }] },
  truncated: { status: "incomplete", incomplete_details: { reason: "max_output_tokens" }, output: [{ type: "message", role: "assistant", content: [{ type: "output_text", text: "Partial answer." }] }] },
};
for (const [name, fixture] of Object.entries(fixtures)) {
  upstreamBody = { id: "resp_fixture", ...fixture, usage: { input_tokens: 1, output_tokens: 2 } };
  for (const stream of [false, true]) {
    const req = new Request("http://localhost/v1/chat/completions", {
      method: "POST", headers: { "content-type": "application/json" },
      body: JSON.stringify({ model: "fixture", messages: [{ role: "user", content: "Hello" }], stream }),
    });
    const res = await handleChatCompletions(req, {} as any, {} as any);
    const text = await res.text();
    if (!stream) {
      console.log(name, "JSON", res.status, JSON.stringify(JSON.parse(text).choices[0]));
    } else {
      const frames = text.split("\n\n").filter(s => s.startsWith("data: {")).map(s => JSON.parse(s.slice(6)));
      console.log(name, "SSE", res.status, JSON.stringify(frames.map(f => f.choices[0])));
    }
  }
}
```

This exercises six handler calls: three identical upstream fixtures, each with the client's `stream` flag false and true. The real request handler returns HTTP 200 for all six. It is a handler-level reproduction with injected upstream responses, not a live-provider/full-server integration test.

### Version

Upstream `dev` commit `a349b521bb02f76c7b8820c13633a8076dcec378`, checked 2026-09-06.

### Operating system

Ubuntu 24.04.3 LTS, Linux x64; Bun 1.4.2.

### Provider and model

Synthetic Responses upstream; model identifier `fixture`. Relevant when the internal Responses replay returns JSON to a streaming Chat request.

### Logs or error output

Normalized observed differences (generated IDs omitted):

| Fixture | Client stream=false | Client stream=true |
| --- | --- | --- |
| Completed function call | `message.tool_calls[0].function.name = "lookup"`; finish `tool_calls` | No `delta.tool_calls`; finish `stop` |
| Reasoning plus text | `reasoning_content = "Fixture reasoning."` and `content = "Answer."` | Text remains; no reasoning delta |
| Token-limited partial answer | `content = "Partial answer."`; finish `length` | Same partial text; finish `stop` |

For the tool fixture, the entire streamed choice sequence is:

```json
[
  {"index":0,"delta":{"role":"assistant","content":""},"finish_reason":null},
  {"index":0,"delta":{},"finish_reason":"stop"}
]
```

It ends with `[DONE]`, without ever transmitting the call ID, function name, or arguments.

### Screenshots and supporting files

Root cause: [src/server/chat-completions.ts, JSON-to-SSE fallback](https://github.com/lidge-jun/opencodex/blob/a349b521bb02f76c7b8820c13633a8076dcec378/src/server/chat-completions.ts#L450-L480).

- The completed Chat object already exists at line 450.
- Lines 462–465 extract only `message.content`.
- Lines 466–470 emit role/text chunks only.
- Line 471 replaces the actual finish reason with the constant `"stop"`.

Suggested fix: serialize the full supported Chat completion message into indexed deltas and reuse the converted choice's finish reason. Keep usage and one terminal `[DONE]`, with the existing stream-budget/lifecycle requirements. Add focused handler tests for tool-only JSON, multiple tool calls, reasoning plus text, and max-output-token truncation.

Distinct from #3767: that report concerns structured refusal content lost inside `src/chat/outbound.ts`. Here the JSON-to-Chat conversion successfully preserves the tool/reasoning fields, and the subsequent SSE synthesis in the HTTP handler discards them.

Duplicate checks included issue searches for JSON fallback/upstream and finish_reason, plus PR searches for JSON fallback, JSON upstream + tool, and synthesized Chat streams. No matching fix/report was found; the defect remains in the pinned current-dev source.

### Redacted configuration

No real configuration is needed. Routing is a synthetic direct Responses fixture in the harness. All identifiers, text, function names, arguments, and usage values are synthetic.

### Checks

- [x] I searched existing issues and documentation.
- [x] I removed secrets, tokens, account details, request credentials, and personal data.

### PR DESCRIPTION
**Pull request by lidge-jun**
## Summary

Preserve tools, reasoning, usage and terminal finish reasons when a streaming Chat request receives Responses JSON. Both native Chat and translated JSON fallbacks use the same serializer. Token-limit and content-filter incompletes retain their own finish reason even when tool output exists; unknown incomplete boundaries return a typed upstream error.

Carry and extend #3779. Closes #3770.

Co-authored-by: Ingwannu <186453546+Ingwannu@users.noreply.github.com>

## Verification

- Independent source reviews passed, including the merge of dev `eff908e0f`. Claude admission/persistence and diagnostic workflows received explicit independent security review. The Windows fixture changes also passed independent source review.
- Local tests, typecheck, installs and builds were not run, per maintainer instruction. Branches were pushed with `--no-verify`.
- [Final integration CI](https://github.com/lidge-jun/opencodex/actions/runs/34058496989) passed all 25 jobs at exact head `4a759716a940595b999805be40d1514c100d63cf`, containing all five layers plus dev `eff908e0f`. All expected jobs completed successfully.
- Dev advanced after final CI started to `f89b815090020f52089801b50f569a777bdfdc0a`. Independent source interaction review passed. That track's [25-job CI](https://github.com/lidge-jun/opencodex/actions/runs/34055675846) verifies its own source; its initial watchdog failure remains recorded there.
- The conflict-free combined source `bf5ce1c3a0ada78f8e506334a82c64b525b859da` additionally passed remote Bun 1.4.0 typecheck, 541 tests across 11 relevant files, and the 425-page documentation build. This supplements the full cross-platform run at `4a759716a`; it is a focused combined-tree check, not a second complete platform matrix.
- Earlier final CI failures were investigated before changing fixtures: [cache-policy controls](https://github.com/lidge-jun/opencodex/actions/runs/34053472964) isolated composed-fixture cold-cache costs; [preparation-budget and semantic-mutation controls](https://github.com/lidge-jun/opencodex/actions/runs/34056824267) verified the competing-OFF allocation and retained assertion sensitivity. [Shim advisory controls](https://github.com/lidge-jun/opencodex/actions/runs/34058337624) additionally verified an owned process deadline while retaining failure on a broken advisory collector. These diagnostics are not full integration proof.
- Automatic PR matrix runs are suppressed by commit skip directives. Lower diagnostics were used only after final-matrix failure. No skipped lower check is described as passed; diagnostic-only workflows are excluded from the delivery stack.

## Checklist

- [x] Scope stays focused and avoids unrelated cleanup.
- [x] Docs or release notes were updated when needed.
- [x] Security-sensitive changes were reviewed for secrets, auth, and unsafe defaults.

**Manual review chain (integrate bottom-up)**

| Layer | PR | Review base |
|---|---|---|
| 1 | #3803 | `dev` |
| 2 | #3804 | `codex/track2-protocol-foundation` |
| 3 | #3805 | `codex/track2-canonical-efforts` |
| 4 | #3806 | `codex/track2-refusal` |
| 5 | #3808 | `codex/track2-claude-compatibility` |

**Maintainer integration decision:** @lidge-jun explicitly integrates this PR into `dev` under MAINTAINERS.md at head `723e59e65814ffac648a9f5ad1043e5c8823d1fd`, using the final-head-first verification above and the authorized admin merge. This is maintainer integration, not a self-approval. Lower skipped checks are not passing evidence. Actor, base, head and review state are refreshed before merging.

**Landing confirmed:** included in dev `5759d9ea2f1e7281cdc01eb9628f2e0a123fb59c`. All five PR heads are ancestors of fetched dev. Final tree `eb4473dfe9e6b95072e3eb31901df635c73dd0eb` exactly equals the remotely checked combined source. #3806 was integrated with #3808; GitHub records both as merged.

### CODE REVIEW

File: src/chat/outbound.ts [hunk 1]
```diff
@@ -560,51 +560,77 @@ export function responsesSseToChatCompletionsSse(
...
+      const call = {
         id: typeof raw.call_id === "string" ? raw.call_id : `call_${uuid().slice(0, 16)}`,
         type: "function",
         function: {
           name: typeof raw.name === "string" ? raw.name : "",
           arguments: typeof raw.arguments === "string" ? raw.arguments : "{}",
         },
-      });
+      };
+      translatorBudget?.chargeRetained(Buffer.byteLength(JSON.stringify(call)), { kind: "retained_collectors" });
```
[the remark this answers is not in the document]
Comment (lidge-jun) [reply src/chat/outbound.ts]: Implemented in 22db6b0a5: complete argument bytes are now reserved under a nonempty internal tool_args scope before serialization, then transferred to aggregate ownership with finally-release. Added exact-limit/over-limit, empty upstream ID, and real handler coverage with valid JSON arguments above 2 MiB. Independent source re-review passed; the updated final integration CI is pending. No local suites were run.

File: src/server/chat-native-sse.ts [hunk 2]
```diff
@@ -94,7 +94,26 @@ export function jsonCompletionSse(value: Rec, requestedModel: string): string {
...
+      const text = `data: ${JSON.stringify(frame)}\n\n`;
+      const bytes = Buffer.byteLength(text);
+      budget?.chargeRetained(bytes, scope);
+      frameBytes += bytes;
+      serialized.push(text);
+    }
+    const done = "data: [DONE]\n\n";
+    const outputBytes = frameBytes + Buffer.byteLength(done);
+    budget?.chargeRetained(outputBytes * 2, scope);
+    return serialized.join("") + done;
```
[the remark this answers is not in the document]
Comment (lidge-jun) [reply src/server/chat-native-sse.ts:113]: Refusal is the separately scoped #3767 change in child PR #3805. That layer populates nullable message.refusal, preserves delta.refusal (including an explicit empty string), reconciles indexed final snapshots, and covers all route/delivery combinations. This foundation addresses #3770 tools/reasoning/finish semantics; the pre-existing refusal gap is not counted as fixed by this layer. Keeping the dependency split rather than duplicating the child implementation here.

**Comment by lidge-jun** [comment]
## 리뷰 · 우선순위 74 / 80

이 PR은 Track2 스택의 맨 아래층이다. 베이스는 현재 `dev`(HEAD `24c761a05`, package 2.45.0)이고, 헤드 브랜치는 `codex/track2-protocol-foundation`이다. 지금 `dev`의 Chat 경로를 보면, 업스트림이 JSON으로 끝난 뒤 스트리밍 클라이언트에게 돌려줄 때 `src/server/chat-completions.ts` 안에서 content만 들고 최소 SSE를 직접 합성한다. tool_calls, reasoning_content, usage, 그리고 incomplete 쪽 finish reason이 스트림에 거의 안 실린다. 같은 의미의 완성본을 이미 `responsesJsonToChatCompletion`이 만들어 두고도, 스트림 배달만 얇게 잘라 버리는 상태다.

이 PR은 그 구멍을 메운다. JSON→Chat 변환에 incomplete reason 분류와 translatorBudget 계측을 넣고, 네이티브·번역 JSON 폴백이 모두 `jsonCompletionSse` 한 길로 나가게 맞춘다. token-limit(`length`)와 content-filter는 도구가 있어도 자기 finish reason을 지키고, 지원하지 않는 incomplete boundary는 typed upstream error로 끊는다. #3779를 이어 받고 #3770을 닫는 층이다. 사용자 눈에 바로 보이는 Chat 스트림 버그이므로 Track2 안에서 가장 먼저 올려야 할 기초다.

스택 위치도 분명하다. #3804(카탈로그 reasoning 정규화), #3805(refusal 보존), #3806(Claude 호환 게이트)가 이 브랜치 위에 쌓여 있다. 최종 통합 CI는 스택 헤드 `84c94f3d77`에서 돌리고, 아래층은 실패할 때만 진단 CI를 요청하는 방식이다. 이전 통합 런(`d60a0716d`)은 25잡 통과했다고 본문에 적혀 있고, 이후 per-call argument cap 보정 커밋이 붙은 뒤 현재 런은 아직 queued다. CI가 초록이 되면 #3803부터 아래로 올리는 순서가 맞다.

코드 쪽 핵심은 `src/chat/outbound.ts`의 `responsesJsonToChatCompletion`과 `src/server/chat-native-sse.ts`의 `jsonCompletionSse`, 그리고 `chat-completions.ts`/`chat-native.ts`의 배달·로그 타이밍이다. JSON projection이 끝나기 전에 200 로그를 남기지 않도록 SSE가 아닐 때는 deferred log를 미루고, 변환/직렬화 실패를 `non_stream` closeReason으로 finalize한다. 테스트는 `tests/responses/chat-json-sse-fallback.test.ts`에 병렬 tool index, reasoning, incomplete 우선순위, 취소, 버퍼 계정, 직렬화 한도를 모아 두었다.

한 가지 범위 메모: 이 PR에 Track2 전체 plan 문서(`devlog/_plan/260907_track2_protocol/*`)가 같이 들어 있다. CodeRabbit도 out-of-scope로 짚었다. 스택 기초 PR에 로드맵을 두는 선택일 수 있지만, 구현 리뷰와 plan 리뷰는 구분해서 보면 된다. refusal 내용이 JSON→SSE에서 빠질 수 있다는 지적도 있었는데, 그건 바로 위층 #3805가 담당한다. 아래층을 먼저 올리는 스택 순서가 그 빈틈을 메운다.

라인 563 근처 src/chat/outbound.ts responsesJsonToChatCompletion - 현재 dev는 incomplete를 전부 length로 뭉개고, 스트림 합성은 content만 내보낸다. 이 PR이 finish reason·tool·reasoning·budget을 채운다
라인 chat-completions.ts JSON stream 합성 블록 - 로컬 최소 SSE를 지우고 jsonCompletionSse 공유 경로로 바꾼 점이 이 PR의 본체다
경로 jsonCompletionSse - tool_calls에 index를 안정적으로 붙이고, 프레임 직렬화 바이트를 budget에 잡는다
경로 chat-native.ts finishLog - 직렬화 성공 전에 200을 찍지 않도록 try 안으로 옮긴 타이밍이 맞다
경로 tests/responses/chat-json-sse-fallback.test.ts - 회귀 범위는 충분해 보이지만, 최종 CI 초록 전에는 “통과”로 말하지 말 것
경로 devlog/_plan/260907_track2_protocol - 구현과 무관한 상위 plan 파일이 같은 PR에 섞여 있다

메인테이너의 판단이 필요한 지점
- Track2 plan 문서를 기초 PR에 같이 둘지, 별도 docs/plan PR로 뺄지
- 최종 CI(`84c94f3d77`)가 아직 queued이므로, 초록 확인 후 merge할지 아니면 아래층 진단 CI도 추가로 돌릴지
- #3805가 refusal을 이어서 고치므로, #3803만 단독 merge해도 사용자 체감 버그의 절반은 남는다 — 스택을 어디까지 한 번에 올릴지

너의 추천
최종 통합 CI가 초록이면 #3803을 맨 먼저 `dev`에 올린다. plan 문서는 남겨도 무방하지만, merge 노트에 “구현 범위는 chat JSON/SSE semantics”라고 짧게 밝혀 두면 좋다. 그다음 #3804→#3805→#3806 순으로 이어서 올린다. 라벨은 건드리지 않는다.

이 댓글은 grok-bot이 작성했습니다

File: tests/responses/chat-json-sse-fallback.test.ts [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+
+interface Chunk {
+  choices: Array<{ index: number; delta: {
+    role?: string; content?: string; reasoning_content?: string;
+    tool_calls?: Array<{ index: number; id: string; type: string; function: { name: string; arguments: string } }>;
+  }; finish_reason: string | null }>;
+  usage?: { prompt_tokens: number; completion_tokens: number };
+}
+
+async function streamFixture(output: unknown[], status = "completed", cancel = false, reason = "max_output_tokens", delivery: { jsonFinish?: string; error?: boolean; errorCode?: string } = {}): Promise<Chunk[]> {
```
[the remark this answers is not in the document]
Comment (lidge-jun) [reply tests/responses/chat-json-sse-fallback.test.ts:21]: The downstream and upstream stream flags intentionally differ. src/server/chat-completions.ts:212 sets internalBody.stream = true for the translated Responses request, then collects it for a JSON Chat client. The upstream assertion must therefore remain true. Both JSON-client boundary cases passed in the current final-head Linux run (34049728209, test 4/4). Changing the assertion to !delivery.jsonFinish would contradict the actual handler and break this regression.

File: tests/responses/chat-json-sse-fallback.test.ts [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+    const rows = getRequestLogEntries().filter(entry => entry.requestId === requestId);
+    expect(rows).toHaveLength(1);
+    expect(rows[0]?.status).toBe(delivery.error ? 502 : 200);
+    const persisted = readUsageEntries().filter(entry => entry.requestId === requestId);
+    expect(persisted).toHaveLength(1);
+    expect(persisted[0]?.status).toBe(delivery.error ? 502 : 200);
+  };
+  if (delivery.error) {
+    expect(response.status).toBe(502);
+    expect(await response.json()).toMatchObject({ error: { type: "upstream_error", code: delivery.errorCode ?? "upstream_incomplete" } });
```
[the remark this answers is not in the document]
Comment (lidge-jun) [reply tests/responses/chat-json-sse-fallback.test.ts:52]: This is a non-blocking coverage-strengthening suggestion rather than a remaining mixed-payload defect. The new overflow path constructs its response through chatCompletionsErrorResponse/chatCompletionsErrorBody, which returns only the error envelope and never copies choices or data. The existing chat-completions-endpoint regression also asserts choices is absent on an error response. The new real-handler regression has passed with HTTP 502, the typed translation_buffer_limit code and exactly one persisted usage row. Retaining that scoped coverage; no production correction is needed for this point.
